Literally a complete rewrite, devouring whole thing with Nix and wg-bond

cab404 opened the pull request:
to:
cc: @subspacecommunity/subspace-maintainers
related to:
resolves:

## Background

Reason for the change: It wasn't working as we wanted it to.

### Changes

* Replace logic module with wg-bond, to support arbitrary network masks
* Replace all the docker with lack of docker


## Testing

We tested it extensively.

cab404 committed:
flake.nix: initial

notgne2 committed:
Fix NixOS module

cab404 committed:
added reload hook

cab404 committed:
fixed service permissions; added reload hooks

notgne2 committed:
Harden systemd service more

cab404 committed:
fixed incorrect admin account redirects

Also deletion messages

notgne2 committed:
fix environment variables for service

notgne2 committed:
use _PREF environment variables for iptables routing in module

notgne2 committed:
add IP addresses in preStart

notgne2 committed:
Add more configurable options to NixOS module

notgne2 committed:
Fix environment values for disableDns

cab404 committed:
initial wg-bond config management

cab404 committed:
flake.nix/systemd: updated teardown/setup process

cab404 committed:
ignoring direnv

cab404 committed:
Merge branch 'master' into wgbond

cab404 committed:
patched wg-quick to skip uid checks

cab404 committed:
Merge pull request #1 from serokell/wgbond

wg-bond configuration management

cab404 committed:
flake.nix: oops

notgne2 committed:
Clean up NixOS module

notgne2 committed:
Remove unused Docker-related files

notgne2 committed:
Restructure flake to use flake-utils and an overlay

notgne2 committed:
Merge branch 'notgne2/cleanup'

agonbar commented:
Hmmm I'd like to ask the rest of the @subspacecommunity/subspace-maintainers about this, as far as I know, most people use this service with docker.

I'd love to add support for Nix, but not at the expense of losing docker.

cab404 commented:
We've PRd this half as joke, and half as a discoverability tool — and also we can build Docker images (requires some more work though) from that thing as well)

gchamon commented:
heard some intresting things about nix, but its nature for system and deployment config is highly experimental AFAIK and the learning curve seems steep. What are the gains of potentally replacing the current set of automation toolings with nix?

notgne2 commented:
NixOS isn't experimental and works just fine in production, but alas you almost definitely _do not_ want it to only work as a NixOS module unless you plan to only use NixOS (and you should of course but we haven't quite taken over the world yet). But as for the packaging with Nix, it is fully reproducible by design, both in the sense of keeping each build pure, but also being able to locally replicate everything, it also means if you don't want Docker, either out of convenience, necessity, or principle, you don't need it, because Nix doesn't suffer from the problems that make Docker useful to begin with for many use cases.

This PR existing is as @cab404 said mostly a joke but if you're actually interested in upstreaming Nix/NixOS support it wouldn't be at the expense of Docker, we just removed it in our fork because we don't intend on ever using it, and it would have needed to be updated to support some changes we made.

We probably would have made a more "compatible" PR for Nix support, if it weren't for us trying to get it to work for us in a hurry (and thus breaking Docker), and the fact that we also gutted and bastardized all of the actual Wireguard code in here in favor of internally using [wg-bond](https://github.com/cab404/wg-bond) to make it work properly for our use cases too, we doubt this would be upstreamed.

If there's any actual interest in Nix support, we can at least upstream this part so so our diff doesn't have to be as big. We could also optionally alter the Docker stuff to just use Nix internally, which would provide some of the benefits of Nix, plus de-duplicate the packaging.

I don't remember the exact issues we had with the existing Wireguard logic, but I remember it being slightly broken, didn't work with some subnets, and didn't work with how we run/orchestrate Wireguard (I think @cab404 remembers more details about this), so we just gutted the whole thing. I think it's unlikely you would be interested in this part, though it has some benefits, so I guess let us know if you are.

rvem committed:
Fix subspace service pre-start script

Problem: Subspace fails to start on a fresh server due to lack of some
configs and directories.

Solution: Create these configs during service pre-start.

rvem committed:
Don't use 'wg syncconf'

Problem: 'wg syncconf' doesn't work as expected and doesn't add routes
for new peers.

Solution: Restart interface via 'wg-quick' instead.

rvem committed:
Merge pull request #3 from serokell/rvem/fix-pre-start-script

Fix subspace service pre-start script

Sereja313 committed:
[Chore] Add systemd unit restart policy

Problem: By default, systemd services generated from the NixOS system configuration don't attempt to restart on failure since Restart=no. However, in some cases, running processes can fail for unclear reasons, and the simplest way to bring the failed service back to life is to restart it. Instead, currently, the service will fail and trigger an alert without attempting to restart.

Solution: Add default values for startLimitBurst, startLimitIntervalSec, Restart, and RestartSec.

Sereja313 committed:
Merge pull request #4 from serokell/sereja/OPS-1402-add-restart-policy

[Chore] Add systemd unit restart policy

Sereja313 committed:
[OPS-1548] Use upstream wg-bond

Problem: wg-bond was removed from recent nixpkgs.

Solution: Use wg-bond from cab404/wg-bond.

Sereja313 committed:
Merge pull request #5 from serokell/sereja/upstream-wg-bond

[OPS-1548] Use upstream wg-bond

cab404 commented:
I like how we're still doing that in that meme PR. @Sereja313 I would advice against using wg-bond from my repo, as it would probably be a subject to a lot of changes, and current implementation in my repo is not compatible with what I've used in this pr at the time

Also haaaai, how's serokell? do you get money for oncalls?)

Sereja313 commented:
Hey, @cab404! 
Serokell is doing quite good. How are you?
> do you get money for oncalls?)


There is currently no strict oncall policy, and offhours work is usually carried over to the next working day. So just paid for working time

> I would advice against using wg-bond from my repo, as it would probably be a subject to a lot of changes, and current implementation in my repo is not compatible with what I've used in this pr at the time

Yea, i see, thank a lot for noticing and sayin it. It stopped building after the last serokell/nixpkgs repin cause wg-bond was removed from upstream nixpkgs earlier this year. So I just pinned it to your last commit without testing it)) I'll pin it to v0.2, thanks again!

cab404 commented:
> There is currently no strict oncall policy, and offhours work is usually carried over to the next working day. So just paid for working time

do demand proper formalized oncall, and payment for that! shit gets stressful)
quite sad that Mostovoy didn't add it after me and the rest of team burned out :(

Sereja313 committed:
[OPS-1548] Pin compatible wg-bond version

Problem: In last commit wg-bond was updated to latest upstream, which
is no lonegr compatible with subspace. See
https://github.com/subspacecommunity/subspace/pull/222#issuecomment-3676753259

Solution: Pin wg-bond to v0.2.0, which was used until it was removed
from nixpkgs.

Sereja313 committed:
Merge pull request #6 from serokell/sereja/upstream-wg-bond

[OPS-1548] Pin compatible wg-bond version
